cxx/pclean/pclean.cc

```diff
@@ -8,12 +8,49 @@
 #include <random>
 
 #include "cxxopts.hpp"
+#include "irm.hh"
 #include "hirm.hh"
 #include "inference.hh"
+#include "util_io.hh"
+#include "pclean/csv.hh"
 #include "pclean/io.hh"
 #include "pclean/schema.hh"
 #include "pclean/schema_helper.hh"
 
+T_observations translate_observations(
+    const DataFrame& df, const T_schema &schema) {
+  T_observations obs;
+
+  for (const auto& col : df.data) {
+    const std::string& col_name = col.first;
+    const T_relation& trel = schema.at(col_name);
+    size_t num_domains;
+    std::visit([&](const auto &r) {
+      num_domains = r.domains.size();
+    }, trel);
+
+    for (size_t i = 0; i < col.second.size(); ++i) {
+      const std::string& val = col.second[i];
+      if (val.empty()) {
+        // Don't incorporate missing values.
+        // TODO(thomaswc): Allow the user to specify other values that mean
+        // missing data.  ("missing", "NA", "nan", etc.).
+        continue;
+      }
+
+      std::vector<std::string> entities;
+      for (size_t j = 0; j < num_domains; ++j) {
+        // Assume that each row of the dataframe is its own entity, *and*
```

Comment:
I don't understand this block of code -- it looks like we're assuming that each entity in each domain has the same (sequential) value for each observation, so the table looks like:
```
D1 D2 D3 val
 0  0  0   x
 1  1  1   y
 2  2  2   z
...
```
is that right? For Model 5, don't we want to read these in from the observations? 

Also, related to the comment, I think we do want to assume that each row of the observation dataframe is its own entity (and is indexed by a primary key domain, like #138 describes), but we don't necessarily want to assume that all ancestor entities are distinct from those of any other entity (adding the index domain should let us avoid that).

Reply:
It is true that in Model 5 we need to assume which entities exist and what their relationships are.  I guess we could create a file with that information, but given that we never know that information "in real life", I think it is better to generate it programatically.  When we get to model 7, we will need to generate the entity assignments programatically anyway (and then allow them to make transitions).

Just to make sure I'm being clear, let me give a concrete example:  the assets/rents_dirty.csv input file, which looks like
Column1,Room Type,Monthly Rent,County,State
0,studio,486.0,Mahoning County,OH
1,4br,2152.0,Clark County,NV
2,1br,1267.0,Gwinnett County,GA
...

There are two classes in the model:  Obs and County.  I agree with you that it is reasonable to assume that each row corresponds to its own Obs entity.  (For now!  For Model 7, we will want the ability to sometimes say that the model thinks that two rows are duplicates of the same underlying entity; that's one of the cleaning operations we want a PClean-type program to be able to do.)  But we aren't given the County entity assignment for each row, and it's not trivial to guess it either, as sometimes either or both of the county name and state fields are missing.  Given that, I think that it is reasonable to initialize the County entity assignment to be unique for each row (which is what my code currently does).

Reply:
I think for `C*`/the "Record" class, we always want to assume that each row is a separate entity and is in one-to-one correspondence with rows of the observed table. A clearer example might be the schema you've defined in other tests, which has as `C*`

```
class Record
  physian ~ Physician
  practice ~ Practice
```

The entities of `Record` are in one-to-one correspondence with the observations, but the physicians and practices they refer to are duplicated.

For testing Model 5, I think (though I'm not sure) that we want to assume we know the entities and their relationships "in real life." This seems important to be able to test that entity clustering is as we expect (we could also hand-build HIRM-like schemas/observations that replicate PClean-style databases, but it would be nice if we could just define PClean-like ones directly). We should probably clarify this with the MIT folks -- I'll post a slack message.

For sampling, eventually we'll want to sample from a County CRP instead of assuming they're unique, using something like the HIRM sampling method introduced in the `080124-emilyaf-sample-hirm` branch (WIP).

Reply:
I've posted about it on slack.

My preference would be to check this code in as-is (or possibly with a TODO or issue to change it depending on the result of the slack conversation).  What do you think?